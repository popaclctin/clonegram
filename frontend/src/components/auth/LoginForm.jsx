import React, { useEffect } from 'react';
import { useLoginMutation } from '../../store/apiSlice';
import { useNavigate, useLocation } from 'react-router-dom';
import { Formik } from 'formik';
import ServerError from '../utils/ServerError';
import './AuthForm.style.scss';

function LoginForm() {
  const navigate = useNavigate();
  const location = useLocation();
  const [loginUser, { isLoading, error, isError, isSuccess }] =
    useLoginMutation();

  useEffect(() => {
    if (isSuccess) {
      const ref = location.state?.from;
      ref
        ? navigate(ref.pathname, { replace: true })
        : navigate('/', { replace: true });
    }
  }, [isSuccess]);

  const validate = (values) => {
    const errors = {};
    //Validate email
    if (!values.email) {
      errors.email = 'Email is required';
    } else if (
      !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)
    ) {
      errors.email = 'Invalid email address';
    }

    //Validate password
    if (!values.password) {
      errors.password = 'Password is required';
    } else if (values.password.length < 6) {
      errors.password = 'Password must be at least 6 characters long';
    }

    return errors;
  };

  return (
    <section className='authForm'>
      <h1 className='authForm__title'>Clonegram</h1>
      <Formik
        initialValues={{ email: 'bula@gmail.com', password: 'parola' }}
        validate={validate}
        onSubmit={(values, { setSubmitting, resetForm }) => {
          loginUser({ email: values.email, password: values.password });
          setSubmitting(false);
        }}
      >
        {(props) => (
          <form onSubmit={props.handleSubmit}>
            <input
              type='email'
              id='email'
              name='email'
              onChange={props.handleChange}
              onBlur={props.handleBlur}
              value={props.values.email}
              placeholder='Email'
              className='authForm__field'
            />
            <input
              type='password'
              id='password'
              name='password'
              onChange={props.handleChange}
              onBlur={props.handleBlur}
              value={props.values.password}
              placeholder='Password'
              className='authForm__field'
            />

            <button
              type='submit'
              className='authForm__submitBtn'
              disabled={!props.isValid}
            >
              Log In
            </button>
            {isError && <ServerError error={error} />}
            {props.touched.email && props.errors.email ? (
              <p className='authForm__error'>{props.errors.email}</p>
            ) : null}
            {props.touched.password && props.errors.password ? (
              <p className='authForm__error'>{props.errors.password}</p>
            ) : null}
          </form>
        )}
      </Formik>
    </section>
  );
}

export default LoginForm;
