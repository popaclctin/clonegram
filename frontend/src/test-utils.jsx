import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from './store';
import { setupListeners } from '@reduxjs/toolkit/dist/query';

import { BrowserRouter } from 'react-router-dom';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
