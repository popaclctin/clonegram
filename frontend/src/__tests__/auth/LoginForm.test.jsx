import React from 'react';
import LoginForm from '../../components/auth/LoginForm';
import { renderWithProviders } from '../../test-utils';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('<LoginForm />', () => {
  it('renders correctly', () => {
    const { asFragment } = renderWithProviders(<LoginForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('logs in the user', async () => {
    const user = userEvent.setup();
    renderWithProviders(<LoginForm />);

    await user.type(screen.getByPlaceholderText(/email/i), 'bula@gmail.com');
    await user.type(screen.getByPlaceholderText(/password/i), 'parola');
    await user.click(screen.getByRole('button', { name: /log in/i }));

    await waitFor(() => {
      expect(
        screen.getByRole('button', { name: /log in/i })
      ).toBeInTheDocument();
    });
  });
});
