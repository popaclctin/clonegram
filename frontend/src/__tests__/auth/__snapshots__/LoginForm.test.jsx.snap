// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> renders correctly 1`] = `
<DocumentFragment>
  <section
    class="authForm"
  >
    <h1
      class="authForm__title"
    >
      Clonegram
    </h1>
    <form>
      <input
        class="authForm__field"
        id="email"
        name="email"
        placeholder="Email"
        type="email"
        value="bula@gmail.com"
      />
      <input
        class="authForm__field"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value="parola"
      />
      <button
        class="authForm__submitBtn"
        type="submit"
      >
        Log In
      </button>
    </form>
  </section>
</DocumentFragment>
`;
